explainaboard/metric.py

```diff
@@ -1,172 +1,258 @@
-        self._results["confidence_score_high"] = confidence_interval_high
+        return MetricResult(self.name, value, conf_interval, conf_value)
 
-        return self._results
+    def get_metadata(self) -> dict:
+        """Return metadata describing the metric in a reproducible way"""
+        return {'name': self.name}
 
 
 class Accuracy(Metric):
-    def __init__(
-        self, true_labels, predicted_labels, is_print_confidence_interval=False
-    ):
-        super(Accuracy, self).__init__()
-        # Metric.__init__(self)
-        self._name = self.__class__.__name__
-        self._true_labels = true_labels
-        self._predicted_labels = predicted_labels
-        self._eval_function = accuracy_score
-        self._is_print_confidence_interval = is_print_confidence_interval
-        self._n_samples = len(self._true_labels)
+    def default_name(cls) -> str:
+        return 'Accuracy'
 
-    def evaluate(self):
+    def calc_stats_from_data(self, true_data: list, pred_data: list) -> np.ndarray:
+        return np.array(
+            [(1.0 if x == y else 0.0) for x, y in zip(true_data, pred_data)]
+        )
 
-        return self._evaluate(self._true_labels, self._predicted_labels)
 
+class F1Score(Metric):
+    def default_name(cls) -> str:
+        return 'F1'
 
-class F1score(Metric):
-    def __init__(
-        self, true_labels, predicted_labels, is_print_confidence_interval=False
-    ):
-        super(F1score, self).__init__()
-        # Metric.__init__(self)
-        self._name = self.__class__.__name__
-        self._true_labels = true_labels
-        self._predicted_labels = predicted_labels
-        self._eval_function = sklearn.metrics.f1_score
-        self._is_print_confidence_interval = is_print_confidence_interval
-        self._n_samples = len(self._true_labels)
+    def __init__(self, average: str = 'micro'):
+        self.average = average
+        supported_averages = ['micro', 'macro']
+        if average not in supported_averages:
+            raise ValueError(f'only {supported_averages} supported for now')
+        super().__init__(name=self.default_name() + average)
 
-    def evaluate(self):
-        # print(self._true_labels[0:10])
-        # print(self._predicted_labels[0:10])
-        # print(sklearn.metrics.f1_score(self._true_labels[0:10], self._predicted_labels[0:10], average='micro'))
-        # exit()
-        return self._evaluate(
-            self._true_labels, self._predicted_labels, average='micro'
-        )
+    def calc_stats_from_data(self, true_data: list, pred_data: list) -> np.ndarray:
+        id_map = {}
+        for word in itertools.chain(true_data, pred_data):
+            if word not in id_map:
+                id_map[word] = len(id_map)
+        n_data = len(true_data)
+        n_classes = 3 * len(id_map)
+        # This is a bit memory inefficient if there's a large number of classes
+        stats = np.zeros((n_data, n_classes * 3))
```

Comment:
not sure if this place is worth adding some comments, such as:
*  `stats[i, id_map[t] * 3]` is used to store the number of actual samples in each class. 
*  `stats[i, id_map[p] * 3 + 1]` is used to store the number of predicted samples in each class.
*  `stats[i, id_map[t] * 3 + 2]`  denotes the number of correctly-predicted samples in class.

Reply:
Added it in the doc string for the whole function.